### PR TITLE
Add stack filters feature

### LINKED ISSUE #57: Enhance find command

### LINKED ISSUE #88: Implement filter function to stack commands

### PR DESCRIPTION
**Pull request by shaunlxw**
Fixes #88.

## Changes
* Update find command to allow find by name, phone, tag, role, and remarks.
* All finds are case-insensitive
* Find by name allows partial match
* Find by phone, tag and role requires exact match
* Find by remarks require full word match that is contained in the remark sentence; finding by empty remark is allowed
* Create Filter class that stores all the filters
* Display the current filters applied after find command
* Add tests for changes

### NetConnect Persons
1. Alex
  * Role: Employee
  * Tags: friends, colleagues
  * Remark: owes me $10
2. Charlotte
  * Role: Client
  * Tags: friends
  * Remark: likes to communicate by email 
3. Roy
  * Role: Supplier
  * Tags:
  * Remark:

### Commands and Results
Command:
```
find n/e
```
Result: alex and charlotte

Command:
```
find n/e
find t/colleagues
```
Result: alex

Command:
```
find t/friends
```
Result: alex and charlotte

Command:
```
find t/frien
```
Result: none

Command:
```
find r/
```
Result: Roy

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -20,15 +28,37 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+
+    /**
+     * Creates the matching NetConnectPredicate based on the provided values in
+     * the given {@code argMultimap}.
+     *
+     * @throws ParseException if the given values are not valid
+     */
+    private static NetConnectPredicate<Person> createPredicate(
+            ArgumentMultimap argMultimap) throws ParseException {
+        if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
```
Comment (purivirakarin) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Shall we handle the case where the user specify both prefix_name and prefix tag?

Currently, if this is the case, the code seems to just filter the search result by PREFIX_NAME.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 2]
```diff
@@ -145,9 +147,22 @@ public ObservableList<Person> getFilteredPersonList() {
...
 
     @Override
-    public void updateFilteredPersonList(Predicate<Person> predicate) {
+    public void clearFilter() {
+        filter = Filter.noFilter();
+        filteredPersons.setPredicate(filter);
+    }
+
+    @Override
+    public void stackFilters(NetConnectPredicate<Person> predicate) {
```
Comment (purivirakarin) [inline src/main/java/seedu/address/model/ModelManager.java]: Nice having this!

**Review by purivirakarin — CHANGES_REQUESTED** [review]
LGTM, just a comment being left on the code. Nice refactoring predicate to support the better filter operation. And nice refactoring the clear predicate / clear filter.

Note: Never looked at the testing code.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java (hunk 1 above)
Comment (shaunlxw) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: This case is handled in line 33 with `ArgumentMultimap#verifyOnlyOnePrefix`! Can be extended to when the find command accepts more arguments too.

**Review by purivirakarin — APPROVED** [review]

**Review by purivirakarin — APPROVED** [review]
LGTM! The find stack for Name Phone Remark Role and Tags work fine.

**Pull request merged**